server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/QueryParser.java

```diff
@@ -0,0 +1,172 @@
+        originMailboxes,
+        subject;
+
+        static FieldName getField(String fieldNameString) throws QueryParserException {
+            return Stream.of(values())
+                .filter(fieldName -> fieldName.name().equals(fieldNameString))
+                .findFirst()
+                .orElseThrow(Throwing.supplier(() -> new QueryParserException("fieldName: '" + fieldNameString + "' is not supported")));
+        }
+    }
+
+    enum Operator {
+        equals,
+        equalsIgnoreCase,
+        contains,
+        containsIgnoreCase,
+        beforeOrEquals,
+        afterOrEquals;
+
+        static Operator getOperator(String operator) throws QueryParserException {
+            return Stream.of(values())
+                .filter(op -> op.name().equals(operator))
+                .findFirst()
+                .orElseThrow(() -> new QueryParserException("operator: '" + operator + "' is not supported"));
+        }
+    }
+
+    interface FieldParser<T> {
+
+        class MailboxIdParser implements FieldParser<MailboxId> {
+
+            final MailboxId.Factory mailboxIdFactory;
+
+            MailboxIdParser(MailboxId.Factory mailboxIdFactory) {
+                this.mailboxIdFactory = mailboxIdFactory;
+            }
+
+            @Override
+            public MailboxId parse(String mailboxIdString) {
+                return mailboxIdFactory.fromString(mailboxIdString);
+            }
+        }
+
+        FieldParser<ZonedDateTime> ZONED_DATE_TIME_PARSER = ZonedDateTime::parse;
+        FieldParser<String> STRING_PARSER = input -> input;
+        FieldParser<Boolean> BOOLEAN_PARSER = Boolean::valueOf;
+        FieldParser<MailAddress> MAIL_ADDRESS_PARSER = mailAddressString -> {
+            return Throwing.supplier(() -> parseMailAddress(mailAddressString)).sneakyThrow().get();
+        };
+
+        static MailAddress parseMailAddress(String mailAddressString) throws QueryParserException {
+            try {
+                return new MailAddress(mailAddressString);
+            } catch (AddressException e) {
+                throw new QueryParserException("mailAddress(" + mailAddressString + ") parsing got error: " + e.getMessage());
+            }
+        }
+
+        T parse(String input);
+    }
+
+    private final ImmutableTable<FieldName, Operator, Function<String, Criterion>> criterionRegistry;
+
+    @VisibleForTesting
+    QueryParser(MailboxId.Factory mailboxIdFactor) {
+        FieldParser.MailboxIdParser mailboxIdParser = new FieldParser.MailboxIdParser(mailboxIdFactor);
+
+        criterionRegistry = criterionRegistryBuilder()
+            .put(originMailboxes, contains, testedValue -> CriterionFactory.containsOriginMailbox(mailboxIdParser.parse(testedValue)))
+            .build();
+    }
+
+    private ImmutableTable.Builder<FieldName, Operator, Function<String, Criterion>> criterionRegistryBuilder() {
+        return ImmutableTable.<FieldName, Operator, Function<String, Criterion>>builder()
```

Comment:
Cardinality still looks allright though I would prefer a more composable approach.

Anyway I agree with what you did.